File: tests/test_pipeline.py
"""
Basic pipeline integration tests.
These test the overall flow and can help students debug issues.
"""
import pytest
import os
import sys

# Add the parent directory to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

def test_pipeline_imports():
    """Test that all pipeline modules can be imported (even if they're stubs)."""
    try:
        import config
        from src.data_loader import load_training_data
        from src.preprocessing import preprocess
        from src.feature_extraction import extract_features
        from src.feature_selection import select_features
        from src.classification import train_classifier
        from src.visualization import visualize_results
        from src.report import generate_report
        from src.utils import save_cache, load_cache
    except ImportError as e:
        pytest.fail(f"Failed to import pipeline modules: {e}")

def test_data_loading_interface():
    """Test that data loading function has correct interface."""
    from src.data_loader import load_training_data
    import config

    # Test function exists and is callable
    assert callable(load_training_data), "load_training_data should be callable"

    # Note: We don't actually call the function here since it might not be implemented
    # Students should expand this test once they implement the function

def test_cache_utilities_interface():
    """Test that cache utility functions have correct interface."""
    from src.utils import save_cache, load_cache

    assert callable(save_cache), "save_cache should be callable"
    assert callable(load_cache), "load_cache should be callable"

@pytest.mark.integration
def test_full_pipeline_structure():
    """
    Test that main.py can be imported without errors.
    This doesn't run the pipeline but checks the structure is valid.
    """
    try:
        import main
        assert callable(main.main), "main.main should be callable"
    except ImportError as e:
        pytest.fail(f"Failed to import main module: {e}")

if __name__ == "__main__":
    # Run tests if script is executed directly
    pytest.main([__file__, "-v"])

File: MATLAB/tests/test_pipeline.m
function test_pipeline()
%% Basic pipeline integration tests for MATLAB implementation.
%% These test the overall flow and can help students debug issues.

fprintf('Running MATLAB pipeline tests...\n');

% Add parent directory to path
addpath('..');
addpath('../src');

% Test 1: Main script can be called without errors (structure test)
try
    % We don't actually run main() since functions might not be implemented
    % Just check that the file exists and can be parsed
    if ~exist('../main.m', 'file')
        error('main.m file not found');
    end
    fprintf('✓ Main script file exists\n');
catch ME
    error('✗ Main script structure test failed: %s', ME.message);
end

% Test 2: Required function files exist in src/
required_functions = {'data_loader.m', 'preprocessing.m', 'feature_extraction.m', ...
                     'feature_selection.m', 'classification.m', 'visualization.m', ...
                     'report.m', 'inference.m', 'utils.m'};

missing_functions = {};
for i = 1:length(required_functions)
    func_file = fullfile('../src', required_functions{i});
    if ~exist(func_file, 'file')
        missing_functions{end+1} = required_functions{i}; %#ok<AGROW>
    end
end

if ~isempty(missing_functions)
    warning('✗ Missing function files: %s', strjoin(missing_functions, ', '));
    fprintf('  Note: Students need to create these function files\n');
else
    fprintf('✓ All required function files exist\n');
end

% Test 3: Cache utilities exist
utils_dir = '../src/utils';
if exist(utils_dir, 'dir')
    if exist(fullfile(utils_dir, 'save_cache.m'), 'file') && ...
       exist(fullfile(utils_dir, 'load_cache.m'), 'file')
        fprintf('✓ Cache utility functions exist\n');
    else
        fprintf('! Cache utility functions missing - students need to implement\n');
    end
else
    fprintf('! Utils directory missing - students need to create\n');
end

% Test 4: Configuration integration
try
    run('../config.m');
    fprintf('✓ Configuration integrates properly with pipeline structure\n');
catch ME
    error('✗ Configuration integration failed: %s', ME.message);
end

fprintf('\nPipeline structure tests completed!\n');
fprintf('Note: Function implementation tests will depend on student code.\n');

end